lib/Language/Haskell/Stylish/Step/Imports.hs

```diff
@@ -160,6 +201,63 @@ formatImports maxCols options moduleStats rawGroup =
   forM_ group \imp -> printQualified options padNames stats imp >> newline
 
 
+--------------------------------------------------------------------------------
+groupAndFormat
+  :: Maybe Int
+  -> Options
+  -> ImportStats
+  -> [NonEmpty (GHC.LImportDecl GHC.GhcPs)]
+  -> Editor.Edits
+groupAndFormat maxCols options moduleStats groups = edits
+  where edits = Editor.changeLines block (const regroupedLines)
+        regroupedLines =
+          intercalate [""] $
+          map (formatImports maxCols options moduleStats) grouped
+
+        grouped = groupByPatterns (groupPatterns options) imports
+        imports = concatMap toList groups
+
+        block = Block
+          (GHC.srcSpanStartLine . src $ head imports)
+          (GHC.srcSpanEndLine   . src $ last imports)
+        src = fromMaybe (error "regroupImports: missing location") .
+          GHC.srcSpanToRealSrcSpan . GHC.getLocA
+
+groupByPatterns
+  :: [Pattern]
+  -> [GHC.LImportDecl GHC.GhcPs]
+  -> [NonEmpty (GHC.LImportDecl GHC.GhcPs)]
+groupByPatterns patterns allImports = go patterns allImports []
+  where
+    go [] [] groups            = groups
+    go [] imports groups       = groups <> [NonEmpty.fromList imports]
+    go (p : ps) imports groups =
+      let
+        (groups', rest) = extract p imports
+      in
+        go ps rest (groups <> groups')
+
+    extract p imports =
+      let
+        (matched, rest) =
+          partition (matches p) imports
+        subgroups =
+          groupBy ((==) `on` capture p) $ sortOn (capture p) matched
+      in
+        (map NonEmpty.fromList subgroups, rest)
+
+    matches :: Pattern -> GHC.LImportDecl GHC.GhcPs -> Bool
+    matches Pattern { regex } import_ = match regex $ moduleName import_
+
+    capture :: Pattern -> GHC.LImportDecl GHC.GhcPs -> String
+    capture Pattern { regex } import_ =
+      case getAllTextSubmatches $ match regex $ moduleName import_ of
+        (_ : s : _) -> map toLower s -- lowercase for case-insensitive sorting
```

Comment:
Well, that answers my earlier question: subsequent matching groups are ignored. I think this should throw an error if there's more than one, better to tell the user they did something unexpected.

Reply:
Just realized that regex-tdfa does not support non-capturing groups at all, which makes the whole capturing thing pretty annoying—there's no way to group stuff without triggering the sub-grouping logic. That could be reason enough to move this to move to [regex-pcre][1] instead. Then using multiple capture groups could result in an error with a message that recommended using non-capturing groups instead.

However, looks like regex-pcre has [an open issue with Unicode support][2], so that isn't great either :/

I'll push some commits for your other suggestions, but going to need a bit more time to think through this capture group stuff.

[1]: https://hackage.haskell.org/package/regex-pcre

[2]: https://github.com/haskell-hvr/regex-pcre/issues/3